pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1150,7 +1157,8 @@ public void run(Timeout timeout) throws Exception {
                 batchMessageAndSend();
             }
             // schedule the next batch message task
-            client.timer().newTimeout(this, conf.getBatchingMaxPublishDelayMicros(), TimeUnit.MICROSECONDS);
+            batchMessageAndSendTimeout = client.timer()
```

Comment:
@zhaijack I think there might still be a small chance to get the exception here. 

Example: 
 * The timeout handler is running
 * We close the producer & cancel the timer
 * We still try to do `newTimeout()` here

I think this should be fixed by putting the `newTimeout()` call inside the above `synchronized (ProducerImpl.this) {}` block. 

That should ensure that either the timer is fired (and then cancelled) or it's not retriggered.

We should also check the producer state in the same sync block. If it's closing/closed we need to ignore the send batch timer and not schedule next timeout.